src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs

```diff
@@ -1882,6 +1883,94 @@ internal override IList<PSModuleInfo> ImportModulesUsingWinCompat(IEnumerable<st
         {
             IList<PSModuleInfo> moduleProxyList = new List<PSModuleInfo>();
 #if !UNIX
+            // the ModuleDeny list is cached in PowerShellConfig object
+            string[] moduleDenyList = PowerShellConfig.Instance.GetWindowsPowerShellCompatibilityModuleDenyList();
+
+            // one of the two parameters can be passed: either ModuleNames (most of the time) or ModuleSpecifications (they are used in different parameter sets)
+            List<string> filteredModuleNames = null;
+            if (moduleNames != null)
+            {
+                if ((moduleDenyList == null) || (moduleDenyList.Length == 0))
+                {
+                    filteredModuleNames = new List<string>(moduleNames);
+                }
+                else
+                {
+                    filteredModuleNames = new List<string>();
+                    foreach (var moduleName in moduleNames)
+                    {
+                        // moduleName can be just a module name and it also can be a full path to psd1 from which we need to extract the module name
+                        var exactModuleName = Path.GetFileNameWithoutExtension(moduleName);
+                        bool match = false;
+                        foreach (var deniedModuleName in moduleDenyList)
+                        {
+                            // use case-insensitive module name comparison
+                            match = exactModuleName.Equals(deniedModuleName, StringComparison.InvariantCultureIgnoreCase);
+                            if (match)
+                            {
+                                string errorMessage = string.Format(CultureInfo.InvariantCulture, Modules.WinCompatModuleInDenyList, exactModuleName);
+                                InvalidOperationException exception = new InvalidOperationException(errorMessage);
+                                ErrorRecord er = new ErrorRecord(exception, "Modules_ModuleInWinCompatDenyList", ErrorCategory.ResourceUnavailable, moduleName);
+                                WriteError(er);
+                                break;
+                            }
+                        }
+
+                        if (!match)
+                        {
+                            filteredModuleNames.Add(moduleName);
+                        }
+                    }
+                }
+
+                // do not setup WinCompat resources if we have no modules to import
+                if (filteredModuleNames.Count == 0)
```

Comment:
I don't think we want to return here, do we?  We want to next check the ModuleSpecification list, and if both lists are empty to then return an empty moduleProxyList?

Reply:
Code had (valid) assumption that only one of the parameters can be passed to it;
But thinking more about - you are right - it is better to move this out and do the check for both lists after both of them are filtered.
Fixed.